docs/design/README.md

```diff
@@ -0,0 +1,595 @@
+- RPM macros.
+- Bootconfig files.
+
+### Docker Domain Socket
+
+Twoliter will mount the host's docker socket (i.e. `/var/run/docker.sock`).
+This means that `docker` commands inside the Twoliter container environment will be interacting with the host's daemon.
+This presents certain challenges because directories need to be mounted from "within" this container.
+In particular:
+
+* The Twoliter container process needs to have `docker` group permissions.
+* The Twoliter container process needs to have permissions to write in certain host directories.
+
+We can solve this by introspecting the Twoliter CLI's process to get the UID and GID it is invoked with.
+We make the assumption that the user invoking the CLI has the permissions to interact with docker and to write in the desired host directories.
+We then pass `--user "uid:gid"` to docker when starting the Twoliter container.
+We can also allow the Twoliter command line to accept uid and gid arguments to override this behavior.
+
+An additional problem to solve is that the paths to mounted directories need to be absolutely the same on the host as they are in the Twoliter container.
+This will not be problem unless the host build or sourcecode directory is in some wildly unusual place, like `/etc/docker` or `/usr/bin`.
+We can prevent this by making a certain set of top-level directories off-limits on the host for the directories we need to mount.
+An error message will describe the problem instead of running the container and failing due to occluded directories.
+
+## Twoliter Projects
+
+Twoliter will have the concept of a project.
+A project can contain multiple variant definitions, multiple kit definitions, and multiple packages.
+Having multiple variants in a single project will help with maintaining similar variants such as
+`my-k8s-1.27` and `my-k8s-1.28`.
+
+A project will have a top level `Twoliter.toml` file (and possibly a `Twoliter.lock` file).
+When `twoliter` commands are invoked, the tool will search for the `Twoliter.toml` file in `.`.
+If it doesn't find it, it will search in `..`, then `../..`, etc.
+This will allow the tool to work from any subdirectory within the project (and is inspired by `cargo`).
+
+`Twoliter.toml` will contain global project settings. Here is a sample `Twoliter.toml` file.
+
+```toml
+# This lets twoliter know if it is reading an older version of the Twoliter.toml
+# schema, or if the directory structure of the project is from an older version.
+schema-version = 1
+
+# This is the name of this project, which can have multiple variants in it. For
+# in-tree builds the project name is "bottlerocket", and thus we produce images
+# with names like `bottlerocket-aws-ecs-1` where the "project" name is a prefix
+# to the variant name. Here we would produce images with names like
+# `ootb-variants-example-dev` where `ootb-variants` is the project name and
+# `example-dev` is the variant name.
+project-name = "ootb-variants"
```

Comment:
I'm not entirely sure about the project name prefix - it seems simpler to just have kits and variants define their own names.

This implicit behavior feels like it complicates the dependency resolution process, because you could copy the same project tree into a different repository and suddenly need to resolve "ootb-variants-core-kit" instead of "bottlerocket-core-kit" because this prefix changed.

Reply:
> I'm not entirely sure about the project name prefix

This was a remnant from earlier thinking, I will delete the explanation and leave it a mystery as to why one might want a name for their project (aesthetic bike shedding?)